### ISSUE TITLE
Dataset processing

### ISSUE THREAD

**Issue by JAMON96** [BPYap/BERTSel#1]
Hello, I was wondering about the part of your program about dataset processing. I saw that the training set of your code contains one question per line, positive_answer, negative_answer. I was wondering how you handled this, taking WikiQA data set as an example

**Comment by BPYap** [comment]
Hi Jamon, for WikiQA, the training set was generated from the `WikiQA-train.txt` file downloaded from https://www.microsoft.com/en-us/download/details.aspx?id=52419. In the original `WikiQA-train.txt`, each training example is given as (question, answer, label) triplet, we just reformat it into (question, positive answer, negative answer) based on the values of the labels.

**Comment by JAMON96** [comment]
Thank you for your reply. I still have a little question to ask you. When you were modifying the WikiQA data set, you modified the data according to its label. I was wondering, in this data set, usually one question corresponds to several answers, did you only take two answers (one positive_answer and one negative answer) in your processing? I wonder if it is convenient for you to share the dataset you processed. Thank you very much for your reply.

**Comment by BPYap** [comment]
The question and positive answer were fixed before considering the negative answers. For each question-positive pair, up to `n` negative answers from that question are selected to form the (question, positive, negative) triplet, however if the number of negatives is less than `n`, we sample the rest from the answers of other questions. At the end, there will be `n` training samples for each (question, positive) pair.

I've uploaded the processed dataset [here](https://entuedu-my.sharepoint.com/:u:/g/personal/boonpeng001_e_ntu_edu_sg/EfK_8F94aCdBpOVrFL9AyUgBgpQsDLNdiWz_4jnxsj4iZw?e=al113B). Hope it helps.

**Comment by JAMON96** [comment]
Thank you very much! I'm sorry that I just saw your PY program about data processing. Thank you for your patient reply, which is very helpful to me！

**Comment by BPYap** [comment]
No problem, glad it helps 😄

**Comment by JAMON96** [comment]
I have successfully completed the training of the data set. Thank you for your patience. I would like to know how you got the TEST_TSV and ANSWER_POOL files in the Inference. Is the inference you are talking about the test stage of the experimental results as I understand it

**Comment by BPYap** [comment]
Yes, the `TEST_TST` and `ANSWER_POOL` files are generated from the test set. Unfortunately, I lost the script that generate those files but I still have the generated files (uploaded [here](https://entuedu-my.sharepoint.com/:f:/g/personal/boonpeng001_e_ntu_edu_sg/EtuC6NQoy21PpuVN6Egfe6cBmUV8udJ6cu5qwsKymz5k-w?e=Xp3kNh)). And if I remember correctly, we were not able to replicate the WikiQA results reported in the BERTSel paper, you might need to check with the original authors regarding the training details.

**Comment by JAMON96** [comment]
When epoch=5, the result of my experiment was better than that of the paper. I don't know if it is because I adjusted batch_size. Under your configuration, the memory of the GPU in my lab is insufficient.
![image](https://user-images.githubusercontent.com/42796458/119936169-4c9bca80-bfbb-11eb-8649-447aa6528de6.png)

**Comment by BPYap** [comment]
Cool! Good to know changing batch size does the trick 😄

**Comment by JAMON96** [comment]
I can't do this without your help. I would like to know if you have any other data sets that you have processed, and I would like to verify that the adjustment of batch_size works under the other data sets

**Comment by BPYap** [comment]
Yes we also tested on the TrecQA dataset, I've updated the cloud storage [link](https://entuedu-my.sharepoint.com/:f:/g/personal/boonpeng001_e_ntu_edu_sg/EtuC6NQoy21PpuVN6Egfe6cBmUV8udJ6cu5qwsKymz5k-w?e=Xp3kNh) with the processed files. Enjoy! :)

**Comment by JAMON96** [comment]
thank  you！

**Comment by JAMON96** [comment]
I also have a small question to consult you. I noticed that there is a part about evaluation in the code, but I did not find the validation set in the data file you gave me. I don't know if you did the verification experiment at the beginning of the experiment. Or do you think the validation task is a must.
![image](https://user-images.githubusercontent.com/42796458/120569022-b0a11180-c447-11eb-99ca-1288a1ac6b1a.png)

**Comment by BPYap** [comment]
The evaluation part is optional, its mainly for tracking the changes in validation score as the training progress. We skipped this step when we're replicating the test set results reported in the paper, but it should be done for best practices.

**Comment by JAMON96** [comment]
Got it. So if I'm going to add validation now, can I make Wikiqa-dev.tsv file format the same as the training set, and then use the validation set to adjust the model?
The comments in this part of your code may not have changed
![image](https://user-images.githubusercontent.com/42796458/120620082-31373080-c48f-11eb-9c61-a61f9b14c56d.png)

**Comment by BPYap** [comment]
Yes you should be able to generate the validation file using the `generate_training.py` script without modifying any code. After that, pass the path of the generated file to `--dev_tsv`, enable `--do_eval` and optionally the `--evaluate_during_training` flag if you want to track the validation score at every logging step.

You're right about the typos in the help description for `--dev_tsv` (and `--train_tsv`). I've changed it to reflect the correct description. Thanks!

**Comment by JAMON96** [comment]
thank you

**Comment by JAMON96** [comment]
As you say, I added the validation set to the experiment, high precision of validation set at the beginning, but with the experiment, the precision of the validation set began to fall, I repeated the test, whether training 1 epoch or 5 epochs, acc to the final results of validation set is very low, this leads me finally the result of the test set is bad, also don't know if you encounter this problem.
As you can see from these graphs, the acc value at the beginning of the training is higher than the value at the end of the training. In addition, the final results of the experiment were poor.
![image](https://user-images.githubusercontent.com/42796458/120923218-f2f47800-c6ff-11eb-8e40-e874885b6fc5.png)
![image](https://user-images.githubusercontent.com/42796458/120923221-f7b92c00-c6ff-11eb-9cea-4ba136da31d4.png)
![image](https://user-images.githubusercontent.com/42796458/120923225-fdaf0d00-c6ff-11eb-930d-ac09729534f8.png)
thank you

**Comment by BPYap** [comment]
Can you post the contents of `training_args.bin` for this experiment? The contents can be extracted with `torch.load`. I'm not sure what went wrong as we have not encounter this before, which dataset were you training on?

**Comment by JAMON96** [comment]
Yes, I did the work on WikiQA data set, training_args.bin file I uploaded [here](https://1drv.ms/u/s!As6WuC8NGmlajwJjdGZh9Gycnrq2?e=bw9gH3), you can have a look.

**Comment by BPYap** [comment]
The training arguments seem fine to me. Perhaps you can try to reduce the batch size by half and add some warmup steps.

> When epoch=5, the result of my experiment was better than that of the paper. I don't know if it is because I adjusted batch_size. Under your configuration, the memory of the GPU in my lab is insufficient.
> ![image](https://user-images.githubusercontent.com/42796458/119936169-4c9bca80-bfbb-11eb-8649-447aa6528de6.png)

Also, are you using the same setups as this experiment above?

**Comment by JAMON96** [comment]
I've increased the batch_size here compared to the first experiment, so let me do a couple of comparison experiments. Thank you
